fullstack/src/components/widget.jsx

```diff
@@ -1,70 +1,99 @@
 import React, { useState } from "react";
-import { Button } from "@/components/ui/button";
+import { useQuery } from "@tanstack/react-query";
 
 // Sensor Mapping
 const sensorMapping = {
   temperature: { unit: "°C", label: "Temperature" },
   pressure: { unit: "hPa", label: "Air Pressure" },
-  humidity: { unit: "%" , label: "Humidity" },
-  windSpeed: { unit: "km/h", label: "Wind Speed" },
+  humidity: { unit: "%", label: "Humidity" },
+  wind: { unit: "km/h", label: "Wind Speed" },
   dust: { unit: "µg/m³", label: "Dust Reading" },
+  co2: { unit: "ppm", label: "CO2 Levels" },
+  gas: { unit: "ppm", label: "Gas Levels" },
 };
 
-const Widget = ({ name, data, GraphComponent, datakey }) => {
+// Fetch data dynamically based on the datakey
+const fetchSensorData = async (dataKey) => {
+  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
+  const response = await fetch(`${baseUrl}/api/${dataKey}-data`);
+
+  if (!response.ok) {
+    throw new Error("Failed to fetch data");
+  }
+
+  return response.json();
+};
+
+const Widget = ({ name, dataKey, GraphComponent }) => {
   const [isExpanded, setIsExpanded] = useState(false);
 
+  // Fetch the data using React Query
+  const { data, error, isLoading } = useQuery({
+    queryKey: [dataKey],
+    queryFn: () => fetchSensorData(dataKey),
+  });
+
   const toggleExpand = () => {
-    if (latestData) { // Only allow toggle if data exists
+    if (data) {
       setIsExpanded(!isExpanded);
     }
   };
 
-  const latestData =
-    Array.isArray(data) && data.length > 0 ? data[data.length - 1] : null;
+  const latestData = data?.length ? data[data.length - 1] : null; // Get the latest data point
```

Comment:
The routes grab the latest one data point at the moment, is this needed?

Reply:
I thought for future we may have the routes fetching an array of more data for displaying on graphs, daily view etc. SO having this means not needing to make an adjustment down the track